nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java

```diff
@@ -0,0 +1,116 @@
+
+@Tags({ "azure", "microsoft", "cloud", "storage", "adlsgen2", "datalake" })
+@CapabilityDescription("Puts content into an Azure Data Lake Storage Gen 2")
+@SeeAlso({})
+@WritesAttributes({ @WritesAttribute(attribute = "azure.filesystem", description = "The name of the Azure File System"),
+        @WritesAttribute(attribute = "azure.filename", description = "The name of the Azure File Name"),
+        @WritesAttribute(attribute = "azure.primaryUri", description = "Primary location for file content"),
+        @WritesAttribute(attribute = "azure.length", description = "Length of the file")})
+@InputRequirement(Requirement.INPUT_REQUIRED)
+
+public class PutAzureDataLakeStorage extends AbstractAzureDataLakeStorageProcessor {
+
+ @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+        final long startNanos = System.nanoTime();
+
+        final String fileSystem = context.getProperty(FILESYSTEM).evaluateAttributeExpressions(flowFile).getValue();
+        final String directory = context.getProperty(DIRECTORY).evaluateAttributeExpressions(flowFile).getValue();
+        final String fileName = context.getProperty(FILE).evaluateAttributeExpressions(flowFile).getValue();
+        final String accountName = context.getProperty(ACCOUNT_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String accountKey = context.getProperty(ACCOUNT_KEY).evaluateAttributeExpressions(flowFile).getValue();
+        final String sasToken = context.getProperty(PROP_SAS_TOKEN).evaluateAttributeExpressions(flowFile).getValue();
+        final String endpoint = String.format(Locale.ROOT, "https://%s.dfs.core.windows.net", accountName);
+        DataLakeServiceClient storageClient;
+                if (StringUtils.isNotBlank(accountKey)) {
+                    final StorageSharedKeyCredential credential = new StorageSharedKeyCredential(accountName,
+                        accountKey);
+                    storageClient = new DataLakeServiceClientBuilder().endpoint(endpoint).credential(credential)
+                        .buildClient();
+                } else if (StringUtils.isNotBlank(sasToken)) {
+                    storageClient = new DataLakeServiceClientBuilder().endpoint(endpoint).sasToken(sasToken)
+                        .buildClient();
+                } else {
+                    throw new IllegalArgumentException(String.format("Either '%s' or '%s' must be defined.",
+                        ACCOUNT_KEY.getDisplayName(), PROP_SAS_TOKEN.getDisplayName()));
+                }
+        final DataLakeFileSystemClient dataLakeFileSystemClient = storageClient.getFileSystemClient(fileSystem);
+        final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);

```

Comment:
Are there any exceptions that can be thrown out of this part of the code and should we be handling the `IllegalArgumentException`? I don't think whether `ACCOUNT_KEY` or `PROP_SAS_TOKEN` are defined will change between retries. It might be worth adding an IT test for that scenario to sort out whether it matters.

Reply:
Thanks @jfrazee , added the IllegalArgumentException to capture the scenario.